### ISSUE TITLE
Missing "Update post" action when there's an autosave

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#16053]
In the Classic Editor, when a published post has an autosave, there are various actions users can take including _update the post_ and get rid of the autosave:

<img width="1279" alt="there is an autosave  classic" src="https://user-images.githubusercontent.com/1682452/59146449-720a2500-89ee-11e9-819f-2916804fcbf6.png">

Instead, in Gutenberg the "Update" button stays disabled:

<img width="1280" alt="there is an autosave gutenberg" src="https://user-images.githubusercontent.com/1682452/59146455-936b1100-89ee-11e9-89b0-504e307a4653.png">

At this point, to get rid of the notice and of the autosave:
- users can try to dismiss the notice, but it will re-appear when opening the post again or when refreshing the page
- users can try to view the autosave but in the Revisions page they can only restore the autosave, there's no option to delete the autosave and save the post "as is"

Instead, in the Classic Editor if users are OK with the post current state, they can just click "Update".

I'd like to suggest to introduce a way to do the same also in Gutenberg.

**Comment by draganescu** [comment]
I second this as the "getting rid" of the autosave is very annoying in the current way.

**Comment by paaljoachim** [comment]
This is a bit more complex. We had a discussion about it during the Gutenberg Design Triage:
https://wordpress.slack.com/archives/C02S78ZAL/p1595953914473900

I am also bringing this issue up during the Open floor for tomorrows Core Editor chat.

**Comment by enriquesanchez** [comment]
I spent some time going over this flow and this is what I found.

## In Gutenberg

As reported in this issue, when there is an autosave on a published post, users are presented with a notice and given the option to **view the autosave**.

At this moment, **updating the post is disabled**.

![Screen Shot 2020-07-30 at 13 50 1](https://user-images.githubusercontent.com/3276087/88981343-549bb880-d28b-11ea-9d47-c23c75d7f1dd.png)

Clicking on **View the autosave** takes the user to the Compare revisions screen. From this screen, users can
only **Restore the autosave** or **Return to the editor.**

![Screen Shot 2020-07-30 at 13 50 2](https://user-images.githubusercontent.com/3276087/88981395-74cb7780-d28b-11ea-9bdb-1782847ca3e6.png)

Clicking on **Restore the autosave** takes the user back to the Editor and **automatically Updates the document**.

![Screen Shot 2020-07-30 at 13 52 1](https://user-images.githubusercontent.com/3276087/88981456-9cbadb00-d28b-11ea-8a8e-9e2613850380.png)

There was no indication or warning that the post was updated. I would've expected to be able to preview the autosaved version in the editor context before being given the choice to update.

That said, If instead, the user selects **Return to the editor**, the autosave notice appears again. The only way to dismiss it is to click on the Close (X) button. However, this does not dismiss the notice permanently. If/when the user returns to the document, they will be presented with this notice again, **until the end of times**, or until new edits are made.

## In the Classic editor

When there is an autosave on a published post, users are presented with a notice and given the option
to **view the autosave**.

Although Updating the post is available, **it won’t update to the autosaved version** (at least not in my testing).

![Screen Shot 2020-07-30 at 17 18 3](https://user-images.githubusercontent.com/3276087/88981765-531ec000-d28c-11ea-9079-90377bb9ba2d.png)

Clicking on **View the autosave** takes the user to the Compare revisions screen. From this screen, users can
only **Restore the autosave** or **Return to the editor**.

![Screen Shot 2020-07-30 at 17 18 2](https://user-images.githubusercontent.com/3276087/88981804-6a5dad80-d28c-11ea-915b-1e340936dcb6.png)

Clicking on **Restore the autosave** takes the user back to the Editor and **automatically Updates the document** (again, I found this unexpected).

![Screen Shot 2020-07-30 at 17 07 1](https://user-images.githubusercontent.com/3276087/88981843-7ea1aa80-d28c-11ea-9783-607770e88f02.png)

But here at least you get a notice of what just happened.

If instead, the user selects **Return to the editor**, the autosave notice appears again. The only way to dismiss it is
to click on the X (close) button. However, this does not dismiss the notice permanently. If/when the user returns to the document, they will be presented with this notice again, **until the end of times**, or until new edits are made.

---

I think there are clear areas of opportunity in this flow, for both editors. I'll expand more in a following comment to avoid this one from getting too long.

**Comment by enriquesanchez** [comment]
I think this flow could be made more obvious and useful if we give users the ability to take a few actions.

When there is an autosave on a published post, users are presented with a notice and given the following options:

- Preview autosave (only in Gutenberg)
- Compare versions
 -Dismiss permanently
- Close button (X)

**Preview autosave**  (only in Gutenberg)
The autosaved version is previewed in the context of the editor and users should be able to choose between the autosave or the previous version.

**Compare versions**
Users can compare versions in the revisions screen. From there, one should be able to restore to the autosave version (as it currently works), but also have the additional option to dismiss/delete the autosave version. If they choose to delete it, they return to the previous version they had in the editor and the notice is gone forever.

**Dismiss/delete the autosave**
Both the notice and the autosave are gone.

**Close button (X)**
Maintains current behavior. Next time the post is opened in the editor, the user is prompted again until they select an option from the above.

---

Thoughts?

**Comment by MDWolinski** [comment]
@enriquesanchez agree with your analysis of this issue.

The only other potential addition would be if the user should also have the ability to remove the newer version from the notice bar so a user wouldn't need to go into the comparison and then delete the autosave.

**Comment by estelaris** [comment]
I would add another clarification to the workflow when comparing revisions. In both of @enriquesanchez 's examples above, in the columns, the different content is identified with red and green. That to users may say what is wrong/right but it doesn't differ the autosave from the "other" version. I am not very technical and I am having issues to figure out which revision should I keep.

![image](https://user-images.githubusercontent.com/11415402/89033311-cda41a00-d336-11ea-94a7-ccbde443214d.png)

**Comment by enriquesanchez** [comment]
Here's a new proposed flow to handle the autosaves. 

When there's an autosave, users will be presented with a notice, similar to how it currently works. This notice will now provide a couple of options: **View autosave** and **Compare versions**. 

**View autosave** will let users see the autosaved version in the context of the editor. Once there, they can **Restore** or **Delete** the autosave.

▶️ [Play with the prototype](https://www.figma.com/proto/5bhvsEg2LxSGXI2Bv7hWR0/Enrique-s-Masterpieces?node-id=449%3A194&viewport=1170%2C-21900%2C3.9135591983795166&scaling=min-zoom)

### View and restore

![view_and_restore](https://user-images.githubusercontent.com/3276087/89473565-ce7ae880-d748-11ea-8563-8942ae758598.gif)


### View and delete

![view_and_delete](https://user-images.githubusercontent.com/3276087/89473573-d5096000-d748-11ea-8dd2-259a69a9509d.gif)

Notice the **'Update'** button being disabled until the user makes a choice.

---

The other option, **Compare versions**, takes the user to the 'Revisions' screen in wp-admin (the current flow). In that screen, the user now has the option to both **Restore** and **Delete** the autosave.

### Compare and restore

![compare_and_restore](https://user-images.githubusercontent.com/3276087/89473750-4a753080-d749-11ea-9f41-5954c0792f5e.gif)

### Compare and delete

![compare_and_delete](https://user-images.githubusercontent.com/3276087/89473759-506b1180-d749-11ea-9658-41de8334f75f.gif)

---

Clicking on the **Close (X)** button will maintain the current flow, where the notice is only dismissed but the autosave flag remains, so next time they open the post, they will see the notice again until they make a choice.

## Feedback  📣
- How does this flow feel in general?
- Do you feel that you know what's happening on every step?
- What do you think of the notices? Do they feel like the right UI for this flow?

**Comment by annezazu** [comment]
I'm not a designer but thought I'd give some feedback in case it's helpful :)

**How does this flow feel in general?**

This feels a ton better than the current version. Less technical and more robust. I wonder if the word "recover" might be better than "restore"? I don't feel strongly but just a thought as I think of restoring as being a more intense action (like restoring from a backup). I worry that might confuse people.

**Do you feel that you know what's happening on every step?**

Yes! Although I found myself wanting in the "View and restore" GIF to feel nervous about deleting. Delete is such a strong word and I feel like it might benefit from having the "Compare Versions" options there as well or some way to get back to the original. Not a dealbreaker but just a reaction I had. In general, I found having the delete option there made my mind go to waiting to have the option to keep it too 🙃. The human brain is an odd thing. 

The other flows look great to me and make sense!

**Comment by enriquesanchez** [comment]
Thanks @annezazu, this is great feedback!

>I wonder if the word "recover" might be better than "restore"?

I love "recover", yes I think it's better choice.

>Delete is such a strong word 

How about "Discard" instead?

>Although I found myself wanting a "Compare Versions" option in the "View and restore" GIF as I felt nervous about deleting.

We wanted to avoid having too many options presented to the user at once, that's why we kept them to only two per step. That said, maybe a way to go back to the first step give users more peace of mind?

**Comment by annezazu** [comment]
Discard sounds great. I don't know how it would hold up when it comes to internationalization but it _feels_ better and less severe. 

> We wanted to avoid having too many options presented to the user at once, that's why we kept them to only two per step. That said, maybe a way to go back to the first step give users more peace of mind?

Makes sense! Ideally, this would be an option but I understand the current flow doesn't have that either so it would be asking for a pretty big change. 

All in all, super excited to see this coming together particularly as I just ran into the current flow earlier today 😆

**Comment by enriquesanchez** [comment]
Here's a revised prototype of the same flow described above. I've update the visuals and took @annezazu's feedback into consideration (thank you!).

I decided to try a different, but similar UI for this flow. I feel that it was putting to much weight on the notices UI, which was not designed and is not intended for this use case. This new UI is quite similar but allows us to break away from any constraints the notices UI may come with.

Only showing one part of the flow. Let me know what y'all think 🙌

![2020-08-07 17-17-00 2020-08-07 17_19_50](https://user-images.githubusercontent.com/3276087/89692936-ab7c4000-d8d2-11ea-8e66-ea5b0a194ca5.gif)

**Comment by draganescu** [comment]
@enriquesanchez to make sure:

- if there is an autosave
- and I get the message
- then click on view autosave
- I get to see the version in the autosave
- and I don't want to neither recover, nor discard the autosave at this point
- to get back to the previous state
- I should click the (X) and dismiss the autosave notice?

**Comment by enriquesanchez** [comment]
>to get back to the previous state 
>I should click the (X) and dismiss the autosave notice?

@draganescu Thanks for the feedback!

@annezazu had the same great observation. I initially didn't consider it but will definitively revise the flow so there's a way to take a step back.

I don't want to add another option to the main actions, but perhaps the "back button" can be part of the component's UI:

<img width="872" alt="noticemodalbar" src="https://user-images.githubusercontent.com/3276087/89928602-af0f1000-dbcd-11ea-8fa7-d5fa9e302705.png">

**Comment by paaljoachim** [comment]
Taking a step back....
The problem with autosave is all the selections that we suddenly have. It can easily create confusion. The easiest choice for a user is to just disregard and keep what is seen.

The existing approach as is seen in WP 5.5 and Gutenberg 8.7.1.

<img width="1326" alt="Screen Shot 2020-08-12 at 12 42 31" src="https://user-images.githubusercontent.com/5323259/90006691-abe15600-dc99-11ea-92f6-03101d840ad8.png">

There is the View the autosave text link and the x icon. 
One can disregard the notice and keep working and click update button to save, but the notice is still there until the user chooses an action. If the user continues to work on the current layout then the notice should automatically be removed.

-----

From the above visual example from @enriquesanchez 
Recover this autosave - Discard this autosave - X.
It seems the Discard this autosave would have the same result as clicking the X. 

-----

Bottom line.
Autosave shows up. User just ignores it and continues to work on the layout. Adjusting and clicking Update button to save the content. Then the Autosave notice should automatically go away.

Autosave can very easily cross over into version/history.

**Comment by enriquesanchez** [comment]
Here's another round of iteration, this time I'm exploring a flow where the user will "enter" into an autosave step by removing the top toolbar UI and instead we retain the user focus on the task at hand.

![2020-08-12 10-42-22 2020-08-12 10_46_15](https://user-images.githubusercontent.com/3276087/90039874-da0c6900-dc8c-11ea-9aa1-1f246e1d988a.gif)

Worth noting:

1. The user can still "exist" this step by clicking on the X button.
2. The new "autosave topbar" provides info on the autosave (by who and when)
3. By offering the ability to preview and compare side-by-side within the editor, we avoid having the user go to another screen/UI in wp-admin.
4. The user can choose between Recover and Discard. Recover restores the autosave. Discard deletes the autosave. Either way, once the user selects one of these options the autosave notice will go away.

## I need your feedback 📣

1. Compared to previous iterations (see explorations above in this issue), how does this "enter into autosave mode" feel to you? Do you have enough context and info to know what to do? If you had second thoughts, would you know how to step back?
2. Do you think we should still offer the option for the user to go to the diff (revisions) screen in wp-admin? If so, when and where would you expect to see this option?
3. What do you think of the side-by-side preview? Is it clear what is current and what is the autosave?

**Comment by mapk** [comment]
This is looking amazing! I've always had problems with this notification, and your new design feels like a great balance between notifying and providing visual context to the decision.

To address your questions:

1. **Entering into autosave mode:** Based on some earlier feedback, maybe try moving the yellow notification (redesigned) into the topbar itself. Right now, with the yellow notification below the topbar, it looks like the user can still interact with certain elements there. But maybe, from the gitgo, this notification should just take over the topbar and require a decision before any editing can happen.
2. I feel like this notification is about a previous auto-save and not about every possible revision. So maybe separating the revision history from this auto-save mode is okay.
3. The side-by-side preview is beautiful. The grey line separating both previews works well. 👍  And I like the clear labels at the top.

**Comment by MichaelArestad** [comment]
@enriquesanchez this is getting interesting as it's getting into visual diffs. When you enter the "compare" mode, it's no longer clear to me which side the Recover/Discard buttons are referring to.

I almost wonder if the entire top toolbar should "split" and have the author/label and the action above each side. Something like this perhaps (needs refinement for sure):

<img width="1157" alt="image" src="https://user-images.githubusercontent.com/1123119/90044624-1d65d800-dc8b-11ea-98a2-cc5cd571db9d.png">

**Comment by joanrho** [comment]
@enriquesanchez Really nice explorations! Just sharing some thoughts here:

- This visual side-by-side comparison of Current v Autosave versions of the page/post is very helpful. The "Current"/"Autosave" labels in each view are helpful, and this might be a good opportunity to align those styles with @dubielzyk on his template part labeling work. I don't think the single Autosave view is necessary here since the comparison is more useful.
- Can we bring back the code view here too so users can just toggle between visual/code views while in this autosave isolation mode?
- I think we should make the CTA buttons "Recover"/"Discard" clearer as to what users can expect will happen when they click each, since Recovering autosave or Discarding autosave feels a bit confusing. Similar to what Michael suggested above, what do you think about buttons that say something like "Keep Current" and "Use autosave"?

**Comment by SNaushadS** [comment]
in context of the autosave, i have some confusion. Probably its only me… Here it is
In the comparison, we are showing _Current | Autosave_
The latest one being towards the right makes sense that it should be the latest version…. however… i got confused with only the wordings…
Is Autosave, more current than the Current ?

if others too feel this can be improved here is a suggestion:
Having a timestamp with the Current (DD-MM-YY, HH-MM-SS) | and Autosave (DD-MM-YY, HH-MM-SS)

**Comment by paaljoachim** [comment]
I really like your latest version @enriquesanchez !
It is wonderful to have both versions side by side.
The wording "Use this version" and "Recover" could be more like @SNaushadS is hinting at.

Kinda like: use version made 2 hours ago/or timestamp or use current version.

As it gives a time stamp to when the older version was made in comparison with the version that is used in the editor.

Btw
I do feel this method of changing/removing the top bar to place focus on choosing the autosave or the current version is a method that can also be used for other features (which features I am not sure off at this moment). As it is an interesting way to guide the user into taking a decision. Perhaps this is a method to be used for a section in Full Site Editing....

**Comment by enriquesanchez** [comment]
Thanks for the great feedback everyone!

I worked a little more on the flow and here's the result:

![2020-08-12 18-15-50 2020-08-12 18_19_01](https://user-images.githubusercontent.com/3276087/90078088-663c8180-dcc9-11ea-8708-03184189153b.gif)

1. After the user clicks on "View autosave", the split-screen is now shown by default with both the Working and Autosave versions side-by-side. 
2. I removed the user avatars with the intent of reducing complexity and keeping just the information that is truly necessary. The user name is still shown along with the timestamp.
3. I added visual indicators in the split line to make it easier for users to spot where the differences are.
4. There are now two buttons, one for each version. 

---


>in context of the autosave, i have some confusion.

@SNaushadS @paaljoachim Is it now clearer which version is which?

>Can we bring back the code view here too so users can just toggle between visual/code views while in this autosave isolation mode?

@joanrho I had a little trouble figuring out a good place for this option, but I agree some users are used to it and even rely on it for certain flows. Where would you expect to find this option?

>I think we should make the CTA buttons "Recover"/"Discard" clearer 

How do you feel about the new button labels?

--- 

**A few things I'm still uncertain about:**

1. Maybe the user doesn't want to deal with this decision right now and wishes to shelve it for later. How would they exit this flow? 
2. How do we translate this screen to mobile and smaller screens where a side-by-side view wouldn't be possible?

**Comment by paaljoachim** [comment]
1. Use case: I see the autosave notification show up but choose to disregard it and continue working on the current version. After having clicked update/save then the autosave notification should automatically go away. As I am in a way saying that I choose to use the current version and have made adjustments and even saved it. That I choose not to relate to an autosave version. The autosave version goes into the revision/history which if needed I can get back by clicking revisions.

2. Mobile. I am wondering if a toggle between current/autosave is needed. So the user can choose to jump between the two. I am also wondering if comparing current and autosave should be more similar to comparing revisions.

Comparing current <-> autosave is somewhat similar to comparing revisions.
Which brings up the questions....
Should clicking autosave bring up the revisions screen? Where we see the code. One side in pink and the other side in green. Or having a way to go between the split screen and the revisions screen?


Btw
This issue is somewhat associated: https://github.com/WordPress/gutenberg/issues/20957
"No visible timestamp for autosave."

**Comment by SNaushadS** [comment]
@enriquesanchez Thanks for the iteration. This appears much cleaner and crisp. The visual weight is also low now, as you've removed the avatar and used secondary buttons.

**Comment by MDWolinski** [comment]
So we're understanding of the flow of where this issue might come up, from my understanding, the user was editing a blog post that's saved in the local browser cache, but closes the browser before an autosave occurs.  When the user returns to edit their post, the local version is different from the last autosave version, thus prompting this issue.

Right?

Does this prompt then stop an autosave from occurring until I either dismiss the alert?  Or if I let my browser sit long enough will an autosave occur and make this issue...er...go away anyway?

Essentially, clicking the X is equivalent to choosing to use the local copy, which when chosen, should probably then initiate an autosave so the local copy/autosave are in sync.

This is separate from revisions, in my mind because the autosave is not necessarily a revision.  If the user has activated revisions (and can we talk about the inability of a user to activate revisions without modifying their wp-config or theme file, just seems like there'd be an easier way for avg users to manage this or not, off soapbox) then a revision is created when the user clicks Update/Post button.

So in this workflow, there is a Saved post, the user is actively editing the document and Local Cache / Autosave is the interim state until the post is saved via Update / Post, thus creating a new revision.

As we talked about in Slack, I think this interface could be used to easily navigate between revisions, my team built a similar interface for one of our corporate sites to do that within the custom editor we use, I'll attach a screenshot below as a sample of this, but overall, this is probably an additional issue that references this work.

**Mobile Version, my thoughts**
And since Enrique did a great implementation of my idea of using the divider line to highlight the differences (but I wonder if the line should be different from grey to just be clearer), I'll throw out a more radial idea for handling this on mobile.

What if, on mobile (small screens), the two versions are overlaid on each other with a slider to transitions from one to the other using transparency?  Could even technically be done on large screens if people wanted, I guess.  So with this interface, I could see the Local version at 30% transparency over the Autosave version and scroll through it to see the differences (the indicator of difference could be on the left or right margin depending on text flow).

**Our Custom Revision Navigator**

So at one of our corporate sites, this is the revision navigator we built...

<img width="1535" alt="Guide_Editor___Financial_Systems_Help" src="https://user-images.githubusercontent.com/22196395/90128394-d16f6d80-dd34-11ea-947d-f1e54a6be873.png">

It does function a little similar to the WP revision viewer, but because we work in a multi-user workflow, the navigator shows who created the revision and the date it was said.  The star shows the current version, the blue shows the selected revision.  If the revision is the same as the current version, then it indicates that there isn't a difference between the two selected versions.

Not saying this is perfect, etc, but as mentioned above, I can certainly see this issue being added on to include revision navigation as well.

**Comment by paaljoachim** [comment]
I have not read through everything again and might have missed something that I could have addressed....

Autosave is just that the post or page automatically saves. 
For instance Google docs has an auto save and various software automatically saves the work one is doing without any user notification. 

I suggest that we just incorporate Autosave into the regular save flow, and that no notification is seen at all for Autosaves. The user can instead just see something like this mockup while editing the page/post.

![Auto-Save-incorporated-into-save-Gutenberg](https://user-images.githubusercontent.com/5323259/92991139-e2e99800-f4e1-11ea-9705-2204eaccf618.jpg)

Today when we click the update button one sees this:
![Screen Shot 2020-09-12 at 09 47 08](https://user-images.githubusercontent.com/5323259/92991129-c5b4c980-f4e1-11ea-8d64-f34dc7c2d182.png)

Since Autosaving is already a part of WordPress I do think it is better to just incorporate it as part of the saving procedure. Be it manual or automatic.

----

Much of what is being discussed above here really belongs to the revisions panel, and not a part of the auto save procedure. If a user does not agree on the point of where the post/page was saved then there are two  options:

Undo/redo to arrive at a point they are satisfied with. Then clicking update/save.
![Screen Shot 2020-09-12 at 10 28 10](https://user-images.githubusercontent.com/5323259/92991229-b4b88800-f4e2-11ea-81f1-9b573fa9e8bc.png)

Or use the revisions panel to compare the code of one version with another.

![Screen Shot 2020-09-12 at 10 29 35](https://user-images.githubusercontent.com/5323259/92991248-e3366300-f4e2-11ea-835f-9f28fcdfa4f3.png)

The revisions panel is something that really needs an update. Using the above comments I am planning on creating a new issue to focus on how we can update revisions so they show up within the Gutenberg area and not exiting the page.

![Screen Shot 2020-09-12 at 10 33 01](https://user-images.githubusercontent.com/5323259/92991314-5b048d80-f4e3-11ea-9c0e-fbcba76b8108.png)

**Comment by paaljoachim** [comment]
I am closing the
"Rethinking the design of the Revisions Panel" issue I made:
https://github.com/WordPress/gutenberg/issues/25270
So that we can keep the discussion going here.

**Comment by paaljoachim** [comment]
I am consolidating another issue:
No visible timestamp for autosave.
https://github.com/WordPress/gutenberg/issues/20957

From support forum: https://wordpress.org/support/topic/autosave-visual-feedback/
My Question is: How (or where) can I see (while writing) when the last Autosave has been executed. In the Classic Editor, there is a Indicator in the Status Bar “Last saved at nn:nn:nn”.

Describe the solution you'd like
A simple way to see when the last autosave was made.

Autosave timestamps can be added into the revisions panel.

**Comment by youknowriad** [comment]
Closing as substantially addressed: the revisions experience this thread consolidated into has been rebuilt — WP 7.1 ships the visual revisions timeline with autosaves explicitly labeled and inspectable, addressing the visibility and recovery concerns gathered here. If the specific restore-notice flow still traps anyone on current versions, a fresh scoped report would be great. Thanks for herding all the related threads over the years! 🙏
